sqldb/invoices.go

```
@@ -0,0 +1,1554 @@
+			Hash:       paymentHash[:],
+			Memo:       sqlStr(string(newInvoice.Memo)),
+			AmountMsat: int64(newInvoice.Terms.Value),
+			// Note: BOLT12 invoices don't have a final cltv delta.
+			CltvDelta: sqlInt32(newInvoice.Terms.FinalCltvDelta),
+			Expiry:    int32(newInvoice.Terms.Expiry),
+			// Note: keysend invoices don't have a payment request.
+			PaymentRequest: sqlStr(string(
+				newInvoice.PaymentRequest),
+			),
+			State:          int16(newInvoice.State),
+			AmountPaidMsat: int64(newInvoice.AmtPaid),
+			IsAmp:          newInvoice.IsAMP(),
+			IsHodl:         newInvoice.HodlInvoice,
+			IsKeysend:      newInvoice.IsKeysend(),
+			CreatedAt:      newInvoice.CreationDate.UTC(),
+		}
+
+		// Some invoices may not have a preimage, like in the case of
+		// HODL invoices.
+		if newInvoice.Terms.PaymentPreimage != nil {
+			preimage := *newInvoice.Terms.PaymentPreimage
+			if preimage == invpkg.UnknownPreimage {
+				return errors.New("cannot use all-zeroes " +
+					"preimage")
+			}
+			params.Preimage = preimage[:]
+		}
+
+		// Some non MPP payments may have the defaulf (invalid) value.
+		if newInvoice.Terms.PaymentAddr != invpkg.BlankPayAddr {
+			params.PaymentAddr = newInvoice.Terms.PaymentAddr[:]
+		}
+
+		var err error
+		invoiceID, err = db.InsertInvoice(ctx, params)
+		if err != nil {
+			return fmt.Errorf("unable to insert invoice: %w", err)
+		}
+
+		// TODO(positiveblue): if invocies do not have custom features
+		// maybe just store the "invoice type" and populate the features
+		// based on that.
+		for feature := range newInvoice.Terms.Features.Features() {
+			params := sqlc.InsertInvoiceFeatureParams{
+				InvoiceID: invoiceID,
+				Feature:   int32(feature),
+			}
+
+			err := db.InsertInvoiceFeature(ctx, params)
+			if err != nil {
+				return fmt.Errorf("unable to insert invoice "+
+					"feature(%v): %w", feature, err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		// Add context to unique constraint errors.
```

I think worth it to copy this over, then can hide all the other errors we add later behind the func: https://github.com/lightninglabs/taproot-assets/blob/main/tapdb/sqlerrors.go#L20-L38